File: Java/Tema7_Ejercicio8.java
/*
 * TEMA 7: FUNCIONES Y MÉTODOS - EJERCICIO 8
 * ==========================================
 * Escriba un programa que use funciones varargs (número variable de argumentos).
 */

public class Tema7_Ejercicio8 {
    
    // Función varargs para sumar números
    static double sumar(double... numeros) {
        double suma = 0;
        for (double num : numeros) {
            suma += num;
        }
        return suma;
    }
    
    // Función varargs para concatenar cadenas
    static String concatenar(String... textos) {
        StringBuilder resultado = new StringBuilder();
        for (String texto : textos) {
            resultado.append(texto);
        }
        return resultado.toString();
    }
    
    // Función varargs para encontrar el máximo
    static double maximo(double... numeros) {
        if (numeros.length == 0) return 0;
        double max = numeros[0];
        for (double num : numeros) {
            if (num > max) max = num;
        }
        return max;
    }
    
    public static void main(String[] args) {
        System.out.println("=== FUNCIONES VARARGS ===");
        
        System.out.println("\n--- Sumar ---");
        System.out.println("sumar(5) = " + sumar(5));
        System.out.println("sumar(5, 10) = " + sumar(5, 10));
        System.out.println("sumar(5, 10, 15, 20) = " + sumar(5, 10, 15, 20));
        
        System.out.println("\n--- Concatenar ---");
        System.out.println("concatenar(\"Java\") = " + concatenar("Java"));
        System.out.println("concatenar(\"Java\", \" es\", \" increíble\") = " + 
                         concatenar("Java", " es", " increíble"));
        
        System.out.println("\n--- Máximo ---");
        System.out.println("maximo(10) = " + maximo(10));
        System.out.println("maximo(45, 23, 78, 12, 89) = " + maximo(45, 23, 78, 12, 89));
    }
}

/*
 * SALIDA ESPERADA:
 * === FUNCIONES VARARGS ===
 * 
 * --- Sumar ---
 * sumar(5) = 5.0
 * sumar(5, 10) = 15.0
 * sumar(5, 10, 15, 20) = 50.0
 * 
 * --- Concatenar ---
 * concatenar("Java") = Java
 * concatenar("Java", " es", " increíble") = Java es increíble
 * 
 * --- Máximo ---
 * maximo(10) = 10.0
 * maximo(45, 23, 78, 12, 89) = 89.0
 */

File: JS/Tema7_Ejercicio8.js
/*
 * TEMA 7: FUNCIONES Y MÉTODOS - EJERCICIO 8
 * =========================================
 * Funciones con argumentos variables (rest parameters).
 * 
 * EXPLICACIÓN:
 * - Rest parameter: ...args
 * - Argumentación variable
 * - Operaciones con múltiples valores
 */

function sumarTodos(...numeros) {
    return numeros.reduce((a, b) => a + b, 0);
}

function concatenarPalabras(...palabras) {
    return palabras.join(" ");
}

function encontrarMaximo(...numeros) {
    return Math.max(...numeros);
}

function filtrarPares(...numeros) {
    return numeros.filter(n => n % 2 === 0);
}

console.log("=== ARGUMENTOS VARIABLES ===");

console.log("Suma de múltiples números:");
console.log(`sumarTodos(1, 2, 3) = ${sumarTodos(1, 2, 3)}`);
console.log(`sumarTodos(10, 20, 30, 40) = ${sumarTodos(10, 20, 30, 40)}`);

console.log("\nConcatenación:");
console.log(`concatenarPalabras("Hola", "mundo", "desde", "JavaScript") = "${concatenarPalabras("Hola", "mundo", "desde", "JavaScript")}"`);

console.log("\nMáximo:");
console.log(`encontrarMaximo(5, 12, 9, 3) = ${encontrarMaximo(5, 12, 9, 3)}`);

console.log("\nFiltrar pares:");
console.log(`filtrarPares(1, 2, 3, 4, 5, 6) = [${filtrarPares(1, 2, 3, 4, 5, 6)}]`);

/*
 * SALIDA ESPERADA:
 * === ARGUMENTOS VARIABLES ===
 * Suma de múltiples números:
 * sumarTodos(1, 2, 3) = 6
 * sumarTodos(10, 20, 30, 40) = 100
 * 
 * Concatenación:
 * concatenarPalabras("Hola", "mundo", "desde", "JavaScript") = "Hola mundo desde JavaScript"
 * 
 * Máximo:
 * encontrarMaximo(5, 12, 9, 3) = 12
 * 
 * Filtrar pares:
 * filtrarPares(1, 2, 3, 4, 5, 6) = [2,4,6]
 */
